From 1b552f09e54466864f246c2078ccce60d7e63314 Mon Sep 17 00:00:00 2001
From: paolo <paolo@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Mon, 22 Jul 2013 15:22:52 +0000
Subject: [PATCH 1/4] 2013-07-22 Paolo Carlini <paolo.carlini@oracle.com>

	PR c++/57920
	* src/c++11/random.cc (random_device::_M_getval): If possible, use
	read instead of std::fread.
	* include/std/random: Do not include <cstdio> unnecessarily.

git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/trunk@201133 138bc75d-0d04-0410-961f-82ee72b054a4

Signed-off-by: Yuanjie Huang <yuanjie.huang@windriver.com>
---
 libstdc++-v3/ChangeLog           | 7 +++++++
 libstdc++-v3/include/std/random  | 1 -
 libstdc++-v3/src/c++11/random.cc | 9 +++++++++
 3 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/libstdc++-v3/ChangeLog b/libstdc++-v3/ChangeLog
index ce67693..c86a805 100644
--- a/libstdc++-v3/ChangeLog
+++ b/libstdc++-v3/ChangeLog
@@ -1,3 +1,10 @@
+2013-07-22  Paolo Carlini  <paolo.carlini@oracle.com>
+
+	PR c++/57920
+	* src/c++11/random.cc (random_device::_M_getval): If possible, use
+	read instead of std::fread.
+	* include/std/random: Do not include <cstdio> unnecessarily.
+
 2014-12-19  Release Manager
 
 	* GCC 4.8.4 released.
diff --git a/libstdc++-v3/include/std/random b/libstdc++-v3/include/std/random
index ceb3825..84b1761 100644
--- a/libstdc++-v3/include/std/random
+++ b/libstdc++-v3/include/std/random
@@ -36,7 +36,6 @@
 #else
 
 #include <cmath>
-#include <cstdio>
 #include <cstdlib>
 #include <string>
 #include <iosfwd>
diff --git a/libstdc++-v3/src/c++11/random.cc b/libstdc++-v3/src/c++11/random.cc
index 2bd7132..1646786 100644
--- a/libstdc++-v3/src/c++11/random.cc
+++ b/libstdc++-v3/src/c++11/random.cc
@@ -30,6 +30,11 @@
 # include <cpuid.h>
 #endif
 
+#include <cstdio>
+
+#ifdef _GLIBCXX_HAVE_UNISTD_H
+# include <unistd.h>
+#endif
 
 namespace std _GLIBCXX_VISIBILITY(default)
 {
@@ -126,8 +131,12 @@ namespace std _GLIBCXX_VISIBILITY(default)
 #endif
 
     result_type __ret;
+#ifdef _GLIBCXX_HAVE_UNISTD_H
+    read(fileno(_M_file), reinterpret_cast<void*>(&__ret), sizeof(result_type));
+#else
     std::fread(reinterpret_cast<void*>(&__ret), sizeof(result_type),
 	       1, _M_file);
+#endif
     return __ret;
   }
 
-- 
1.9.1

